src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -2100,46 +2625,79 @@ UCS_PROFILE_FUNC_ALWAYS(ucs_status_t, uct_ib_mlx5_devx_reg_exported_key,
                         (md, memh), uct_ib_mlx5_md_t *md,
                         uct_ib_mlx5_devx_mem_t *memh)
 {
-    struct mlx5dv_devx_obj *cross_mr;
+    struct mlx5dv_devx_obj *cross_mr = NULL;
+    uct_ib_mlx5_devx_umr_t *umr      = NULL;
+    uct_ib_mlx5_devx_mr_t *mr        = &memh->mrs[UCT_IB_MR_DEFAULT];
     uint32_t exported_lkey;
     ucs_status_t status;
 
     if (memh->dm != NULL) {
         return UCS_ERR_UNSUPPORTED;
     }
 
-    ucs_assertv(memh->cross_mr == NULL,
-                "memh=%p cross_mr=%p exported_lkey=0x%x", memh, memh->cross_mr,
-                memh->exported_lkey);
+    ucs_assertv((memh->umr_mr == NULL) && (memh->cross_mr == NULL),
+                "memh=%p umr_mr=%p cross_mr=%p exported_lkey=0x%x", memh,
+                memh->umr_mr, memh->cross_mr, memh->exported_lkey);
     if (!(md->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI)) {
         goto out_umem_mr;
     }
 
-    ucs_assert(!(memh->super.flags & UCT_IB_MEM_MULTITHREADED));
-    status = uct_ib_mlx5_devx_reg_ksm_data_contig(md, memh->address,
-                                                  (uint64_t)memh->address, 0, 0,
-                                                  "exported-key",
-                                                  &memh->mrs[UCT_IB_MR_DEFAULT],
-                                                  &cross_mr, &exported_lkey);
-    if (status != UCS_OK) {
-        return status;
+    /* Current bind impl (IBV_WR_BIND_MW) attaches a single KLM segment, so:
+     * - IBV_WR_BIND_MW supports the maximum region length of 2GB
+     * - IBV_WR_BIND_MW does not support multi-segment (multi-threaded) MRs
+     * For these use cases we fallback to KSM
+     */
+    if ((mr->super.ib->length > UCT_IB_MD_MAX_MR_SIZE) ||
+        (memh->super.flags & UCT_IB_MEM_MULTITHREADED)) {
+        status = uct_ib_mlx5_devx_reg_ksm_data(md, memh, UCT_IB_MR_DEFAULT, 0,
```

Comment:
why changed uct_ib_mlx5_devx_reg_ksm_data_contig to uct_ib_mlx5_devx_reg_ksm_data?

Reply:
`uct_ib_mlx5_devx_reg_ksm_data` is a helper function that supports both contiguous and MT MRs, while `uct_ib_mlx5_devx_reg_ksm_data_contig` does support only the contiguous MRs. We discussed this with Ivan and Mikhail and agreed that we need to support MT memory handles here as well, like in other places